contracts/eosio.system/src/voting.cpp

```diff
@@ -179,9 +166,10 @@ namespace eosiosystem {
          _gstate.last_producer_schedule_size = static_cast<decltype(_gstate.last_producer_schedule_size)>( producers.size() );
       }
 
-      // Only set a new finalizer policy if it has changed.
-      if( is_savanna_consensus() && !new_finalizer_key_ids.empty() ) {
-         set_finalizers( std::move(next_finalizer_authorities), new_finalizer_key_ids, kept_finalizer_key_ids );
+      // set_proposed_finalizers() checks if last proposed finalizer policy
+      // has not changed, it will not call set_finalizers() host function.
+      if( is_savanna_consensus() ) {
+         set_proposed_finalizers_if_changed( proposed_fin_keys );
```

Comment:
This should just call the new `set_proposed_finalizers`. That function would always check if it changed or not before calling the `set_finalizers` host function rather than trusting the caller to do the right thing. This means a redundant check when called from `actfinkey`, but that's okay.